### PR TITLE
Async options filtering in CustomSelect component

### PR DESCRIPTION
**Pull request by volfkarlo**
# Description
Changelog:
- add processMetadata option to fetch-wp-api
- add filter async options function to custom select

Result:
- ability to adjust, filter, group, etc. dynamically fetched options in custom select
- ability to get additional field(s) from WP API to enrich fetch-wp-api return object

Issue #531

### CODE REVIEW

File: scripts/components/custom-select/custom-select.js [hunk 1]
```diff
@@ -34,6 +34,7 @@ import AsyncSelect from "react-select/async";
  * @param {boolean} [props.hideSelected=false]                         - If set to `true`, the selected option is hidden from the menu.
  * @param {string} [props.loadingMessage='Loading']                    - Text to display when loading options.
  * @param {string} [props.noOptionsMessage='No options']               - Text to display when no options are available.
+ * @param {function} [props.filterAsyncOptions]                        - Function that wraps async options after async fetch. Enables adjusting, filtering, grouping etc. of the fetched data. Return format should match standard select options.
```
Comment (goranalkovic-infinum) [inline scripts/components/custom-select/custom-select.js:37]:
```suggestion
 * @param {function} [props.filterAsyncOptions]                        - Allows modifying (filtering, grouping, ...) options output after the items have been dynamically fetched. Please make sure to include `label` and `value` keys, additional fields can be added as required.
```

File: scripts/editor/fetch-wp-api.js [hunk 2]
```diff
@@ -6,9 +6,10 @@ import apiFetch from '@wordpress/api-fetch';
  *
  * @param {string} endpoint     Endpoint to fetch from (usually post type or taxonomy).
  * @param {object} [options={}] Additional options for fine tunning.
- * 
+ *
  * @param {function} processId              Function that allows custom id processing.
  * @param {function} processLabel           Function that allows custom select option label processing.
+ * @param {function} processMetadata        Function that allows custom field that can be used for aditional filtering/processing.
```
Comment (goranalkovic-infinum) [inline scripts/editor/fetch-wp-api.js:12]:
```suggestion
 * @param {function} processMetadata        Function that allows modifying output data (e.g. for adding metadata for grouping inside a CustomSelect).
```

**Review by goranalkovic-infinum — CHANGES_REQUESTED** [review]
Minor doc comment suggestions. Nice job!

**Review by iruzevic — CHANGES_REQUESTED** [review]
Looks good but please create and example in the readme of this component how this is used

**Review by dingo-d — APPROVED** [review]

**Review by iruzevic — APPROVED** [review]

**Review by dadadavorin — APPROVED** [review]

File: scripts/components/custom-select/docs/readme.mdx [hunk 3]
```diff
@@ -50,6 +50,27 @@ Your IDE will display all the properties, together with their descriptions.
...
+	label={__('My label', 'eigthshift-frontend-libs')}
+	filterAsyncOptions={(options) => !Array.isArray(options) ?
+		options :
+		[...options].map(({label, value, metadata}) => {
+			return {
+				label: `${label} -- ${metadata}`,
+				value,
+			};
+		});
+	}
```
Comment (goranalkovic-infinum) [inline scripts/components/custom-select/docs/readme.mdx:70]:
```suggestion
	filterAsyncOptions={(options) => {
			if (!Array.isArray(options)) {
				return options;
			}

			return [...options].map(({label, value, metadata}) => {
				return {
					label: `${label} -- ${metadata}`,
					value,
				};
			});
		}
	}
```

**Review by goranalkovic-infinum — CHANGES_REQUESTED** [review]
One more minor thing 😄

File: scripts/components/custom-select/docs/readme.mdx (hunk 3 above)
Comment (volfkarlo) [reply scripts/components/custom-select/docs/readme.mdx:70]: I did it on purpose so it's a bit shorter. But if you insist 🙂

**Review by iruzevic — APPROVED** [review]

**Review by dingo-d — APPROVED** [review]

**Review by goranalkovic-infinum — DISMISSED** [review]

**Review by goranalkovic-infinum — APPROVED** [review]

**Review by gabriel-glo — APPROVED** [review]

**Pull request merged**